vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java

```diff
@@ -3300,6 +3300,15 @@ private void confirmUpdate(int id) {
 
     @ClientCallable(DisabledUpdateMode.ALWAYS)
     private void setRequestedRange(int start, int length) {
+        if (length > 500 && length / getPageSize() > 10 && isAllRowsVisible()) {
+            throw new IllegalArgumentException(
+                    "Attempted to fetch more items from server than allowed in one go. "
+                            + "Maximum allowed page count is 10. Consider not using setAllRowsVisible(true) "
+                            + "when you have a large amount of items (not only to cover this issue but also "
+                            + "to avoid performance bottlenecks resulting from rendering thousands of DOM items). "
```

Comment:
Could also mention the data payload transfer as a performance issue, something like: `...performance bottlenecks resulting from transferring the full item data set at once and then rendering an excess amount of DOM elements)...`